Import Nightbot Custom Commands & Timers

redbeardymcgee opened the pull request:
This is a quick and dirty first pass at fetching a [user's custom commands](https://api-docs.nightbot.tv/#get-custom-commands) from their Nightbot.

I ran `prettier` over the file once to reduce some of the line lengths so I could read the file better. There were a couple of `eslint` warnings that I manually fixed up. Apologies for the whitespace churn in this initial commit!

Meaningful changes begin at L#188, but a lot of this will probably be rewritten after the Nightbot OAuth token includes the `commands` scope. I think that happens [here](https://github.com/sogebot/sogebot.dev/blob/a097fcccda05ecd26bd97a8fde2268a12767c2f1/services/credentials/authenticators/nightbot.go#L28C3-L28C3) but I'm pretty sure I can't test a change to that locally since the oauth callbacks won't work on my localhost.

redbeardymcgee committed:
feat(import): rebase onto upstream main

This should include the updated public url unless I messed up bad.

sogehige commented on ui.admin-mui/src/components/Settings/Import/nightbot.tsx:
```diff
@@ -153,44 +170,180 @@ const PageSettingsModulesImportNightbot: React.FC<{
+            variant="contained"
+            disabled={user === 'Not Authorized'}
+            onClick={() => {
+              console.log(fetchCustomCommands());
```
you don't need it here as you are using it inside importCustomCommands

sogehige replied:
```suggestion
```

sogehige commented on ui.admin-mui/src/components/Settings/Import/nightbot.tsx:
```diff
@@ -153,44 +170,180 @@ const PageSettingsModulesImportNightbot: React.FC<{
+            color="primary"
+            variant="contained"
+            disabled={user === 'Not Authorized'}
+            onClick={() => {
```
```suggestion
            onClick={importCustomCommands}
```

sogehige commented on ui.admin-mui/src/components/Settings/Import/nightbot.tsx:
```diff
@@ -153,44 +170,180 @@ const PageSettingsModulesImportNightbot: React.FC<{
+            disabled={user === 'Not Authorized'}
+            onClick={() => {
+              console.log(fetchCustomCommands());
+              return importCustomCommands;
```
```suggestion
```

sogehige commented on ui.admin-mui/src/components/Settings/Import/nightbot.tsx:
```diff
@@ -153,44 +170,180 @@ const PageSettingsModulesImportNightbot: React.FC<{
+            onClick={() => {
+              console.log(fetchCustomCommands());
+              return importCustomCommands;
+            }}
```
```suggestion
```

redbeardymcgee committed:
refactor(import): pull out nightbot types and fns

redbeardymcgee committed:
feat(import): custom command fetching complete

Right now the fetching is finished, but I haven't inspected the api
for creating commands in sogebot.

This commit also factors out the fetching function to reuse its retry
logic in both the command and playlist fetchers.

redbeardymcgee committed:
feat(import): convert nb commands to sogebot

The mapping from the properties of nightbot's commands to sogebot's
properties isn't perfect yet.

I had a different problem trying to delete the commands from
`/commands/customcommands/`. All returned an error, but all of the
imported commands were removed from the UI. IIRC it was a 404, which
I can currently recreate by manually adding any new command and
removing it.

redbeardymcgee commented:
At this point, I have a working version of the full custom command import. Support for special nightbot features (if any such features exist) is not implemented.

However, my local instance of the dashboard returns a 404 when deleting commands. This includes the imported commands, or commands manually created through the UI at `/commands/customcommands`. Works fine with the same instance of my local bot through `dash.sogebot.xyz`, so my local environment has to be broken.

redbeardymcgee committed:
fix(import): don't add any nightbot prefixes

The command group will probably suffice as a way to differentiate
these as external commands. No need to make it explicit thrice per
command.

I need to understand how to map nightbot's userLevel property to
sogebot's permission property on the responses. Also, `order` and
`stopIfExecuted` are currently vague to me.

redbeardymcgee committed:
feat(import): map permissions across bots

redbeardymcgee committed:
fix(import): generate unique ids on import

This is necessary because I could not find what technique Nightbot
uses to generate its command IDs, but even if I did, I think there
is still a slim chance that some ID could clash with an existing
sogebot command ID. I don't understand cryptography enough to
confirm this, but I think the clash is possible because Nightbot
may be using a weaker (older?) algorithm.

Also, Nightbot either doesn't generate a UUID with dashes in it,
or strips them out. Am I supposed to take them out or leave them in?

redbeardymcgee committed:
fix(import): exclude id to allow auto generation

redbeardymcgee commented:
This is, in some sense of the word, complete. However, there are some weird things about Nightbot that I think are missing direct analogues in Sogebot.

First, Nightbot allows commands with no prefix, while Sogebot requires exactly the `!` character to begin any command. It's a fun feature sometimes though. For example, you could set up non-prefixed commands to make the bot help build a `Kappa` pyramid (or to interrupt some emote pyramid for the opposite joke lol). For now, I'm allowing this to fail but I could probably just prefix it with an exclamation point.

Second, I didn't figure out the meaning of the `filter` key on responses. Probably an empty string is fine, but I'd like some confirmation.

Third, Nightbot supports a variety of [special variables](https://docs.nightbot.tv/commands/variableslist) within its commands. [Some of them are not worth porting or converting](https://docs.nightbot.tv/variables/djkhaled), but at least [a few of them](https://docs.nightbot.tv/variables/channel) probably have direct analogues in Sogebot. I'm not sure which ones should be converted to Sogebot's conventions, if any.

Fourth, and I believe *finally*, this feature requires a new scope `commands`. I noted in the original comment on this PR that I think I found where to add the scope, but I can't test it locally to confirm. When I did try to add the scopes [here](https://github.com/sogebot/sogebot.dev/blob/a097fcccda05ecd26bd97a8fde2268a12767c2f1/services/credentials/authenticators/nightbot.go#L28C3-L28C3), the query parameters generated when clicking `Authorize` still only had `scope=song_requests_playlist` without adding `commands`. I guess I didn't make the change correctly, or that's not the right place after all :)

Thanks again for all your patience while I work through this, and for building this cool bot to make it possible for me to do this :)

sogehige commented:
> First, Nightbot allows commands with no prefix, while Sogebot requires exactly the ! character to begin any command. It's a fun feature sometimes though. For example, you could set up non-prefixed commands to make the bot help build a Kappa pyramid (or to interrupt some emote pyramid for the opposite joke lol). For now, I'm allowing this to fail but I could probably just prefix it with an exclamation point.

You can probably use keyword instead of command

> Second, I didn't figure out the meaning of the filter key on responses. Probably an empty string is fine, but I'd like some confirmation.

That's ok

> Third, Nightbot supports a variety of [special variables](https://docs.nightbot.tv/commands/variableslist) within its commands. [Some of them are not worth porting or converting](https://docs.nightbot.tv/variables/djkhaled), but at least [a few of them](https://docs.nightbot.tv/variables/channel) probably have direct analogues in Sogebot. I'm not sure which ones should be converted to Sogebot's conventions, if any.

Hmm, porting some of those may be tricky, although its up to you if you want to do some porting. I would rather add some warning or notification that user should revisit imported commands and adjust them manually for bot.

> Fourth, and I believe finally, this feature requires a new scope commands. I noted in the original comment on this PR that I think I found w[here](https://github.com/sogebot/sogebot.dev/blob/a097fcccda05ecd26bd97a8fde2268a12767c2f1/services/credentials/authenticators/nightbot.go#L28C3-L28C3) to add the scope, but I can't test it locally to confirm. When I did try to add the scopes here, the query parameters generated when clicking Authorize still only had scope=song_requests_playlist without adding commands. I guess I didn't make the change correctly, or that's not the right place after all :)

That's correct, you need to update it there, then I need to rebuild service and deploy it on server.

redbeardymcgee committed:
fix(import): bareword commands become keyword

redbeardymcgee committed:
fix(import): add `commands` scope for nightbot

redbeardymcgee committed:
refactor(import): mostly rearrangement

redbeardymcgee commented:
I don't really think the effort of trying to port Nightbot's special variables is worthwhile until we have a user who depends on it. I just added a snackbar notif for now to go look at the imported commands.

I don't think I have any other notes for this PR then. If no one else has any objections to the code, I'm ready to call this done.

redbeardymcgee committed:
style(import): remove extra linebreaks

redbeardymcgee committed:
feat(import): import nightbot timers

sogehige commented:
Added "commands" scope to service, which should be deployed in a while so you can test it

redbeardymcgee committed:
fix(import): need `timers` scope also

redbeardymcgee commented:
Thanks, the `commands` scope worked but I forgot to add the `timers` scope as well. After that is included, this will be feature complete.

IMO, playlist, custom commands and timers are the main 3 features that Nightbot users actually use a lot. Someone may, someday, request another scope to import but that can be built up in like 30 minutes now that I made 3 examples. 

Ready to merge whenever you say :)

redbeardymcgee committed:
fix(import): don't export unnecessary symbols

sogehige commented on ui.admin-mui/src/components/Settings/Import/nightbot.tsx:
```diff
@@ -45,152 +51,90 @@ const PageSettingsModulesImportNightbot: React.FC<{
+          >
+            Import timers
+          </Button>
         </Stack>
```
Thinking of little bit better styling, wdyt?

```suggestion
       <Stack direction="row" spacing={1} sx={{ pt: 1 }}>
          <Button
            fullWidth
            color="primary"
            variant="contained"
            disabled={user === 'Not Authorized'}
            onClick={() => importPlaylist(accessToken)}
          >
            Import playlist
          </Button>
          <Button
            fullWidth
            color="primary"
            variant="contained"
            disabled={user === 'Not Authorized'}
            onClick={() => importCustomCommands(accessToken)}
          >
            Import commands
          </Button>
          <Button
            fullWidth
            color="primary"
            variant="contained"
            disabled={user === 'Not Authorized'}
            onClick={() => importTimers(accessToken)}
          >
            Import timers
          </Button>
        </Stack>
```

sogehige commented:
Can you comment on my suggestion? Then we can merge :)

sogehige commented:
Timers 

> Thanks, the `commands` scope worked but I forgot to add the `timers` scope as well. After that is included, this will be feature complete.
> 
> IMO, playlist, custom commands and timers are the main 3 features that Nightbot users actually use a lot. Someone may, someday, request another scope to import but that can be built up in like 30 minutes now that I made 3 examples.
> 
> Ready to merge whenever you say :)

`timers` scope added to service

redbeardymcgee committed:
fix(import): improve styling for buttons
